File: orchestrator/repo_env.py
"""Per-repository environments: variables, secret files, services and the checks that prove the work.

Many projects cannot run their tests without configuration that is never committed: a `.env` with
database credentials, an API key, a cookies file. People keep that on their own machine; Relay keeps
it here, per repository, under its data folder, and hands it to every task in that repository:

  * variables   exported to agents, setup, services and verification, and (optionally) written to
                the worktree's `.env` so apps that load dotenv files see them too
  * files       written into the worktree at their relative path and excluded from git
  * system_packages  apt packages Relay installs (once, cached) before setup, e.g. unixodbc for pyodbc
  * setup       replaces the detected dependency install when set
  * services    commands started before the team works (a database, a mock server) and stopped after
  * checks      the commands that prove the work; required ones block delivery, optional ones inform

Secret values never reach the browser after they are saved (the API returns them masked) and are
replaced with ●●●● in logs, messages and artifacts.
"""
from __future__ import annotations

import hashlib
import re
from pathlib import Path

from .util import DATA_DIR, quiet, read_json, write_json

ENV_DIR = DATA_DIR / "repo_env"
MASK = "●●●●"
_NAME = re.compile(r"^[A-Za-z_][A-Za-z0-9_]*$")


def repo_key(repo) -> str:
    """Stable key for a repository: its GitHub name when it has one, else its absolute path."""
    repo = Path(repo).resolve()
    url = (quiet(["git", "remote", "get-url", "origin"], cwd=repo, timeout=15).stdout or "").strip() if repo.exists() else ""
    m = re.search(r"github\.com[:/]+([^/]+/[^/]+?)(?:\.git)?/?$", url)
    ident = m.group(1).lower() if m else str(repo)
    return hashlib.sha1(ident.encode()).hexdigest()[:16]


def source_repo(path) -> Path | None:
    """The repository a worktree belongs to (or the repository itself)."""
    try:
        common = (quiet(["git", "rev-parse", "--path-format=absolute", "--git-common-dir"], cwd=path, timeout=15).stdout or "").strip()
    except Exception:
        return None
    if not common:
        return None
    p = Path(common)
    return p.parent if p.name == ".git" else p


def _path(repo) -> Path:
    return ENV_DIR / f"{repo_key(repo)}.json"


def empty() -> dict:
    return {"vars": [], "files": [], "write_dotenv": True, "setup": "", "services_up": "", "services_down": "", "checks": [],
            "system_packages": []}


def load(repo) -> dict:
    data = read_json(_path(repo), None) or {}
    out = empty()
    out.update({k: v for k, v in data.items() if k in out or k in ("repo", "updated")})
    return out


def _clean_rel(rel: str) -> str:
    rel = str(rel or "").strip().replace("\\", "/").lstrip("/")
    parts = [p for p in rel.split("/") if p not in ("", ".")]
    if not parts or ".." in parts or parts[0] == ".git":
        raise ValueError(f"Invalid file path: {rel!r}")
    return "/".join(parts)


def save(repo, incoming: dict) -> dict:
    """Store what the browser sent. A secret sent back as the mask keeps its stored value."""
    cur = load(repo)
    old_vars = {v["name"]: v for v in cur["vars"]}
    old_files = {f["path"]: f for f in cur["files"]}
    out = empty()
    seen = set()
    for v in incoming.get("vars") or []:
        name = str(v.get("name") or "").strip()
        if not name:
            continue
        if not _NAME.match(name):
            raise ValueError(f"Invalid variable name: {name!r}")
        if name in seen:
            raise ValueError(f"Variable {name} is listed twice")
        seen.add(name)
        value = v.get("value")
        if value == MASK and name in old_vars:
            value = old_vars[name]["value"]
        out["vars"].append({"name": name, "value": "" if value is None else str(value), "secret": bool(v.get("secret"))})
    for f in incoming.get("files") or []:
        if not str(f.get("path") or "").strip():
            continue
        rel = _clean_rel(f["path"])
        content = f.get("content")
        if content == MASK and rel in old_files:
            content = old_files[rel]["content"]
        out["files"].append({"path": rel, "content": str(content or ""), "secret": bool(f.get("secret", True))})
    out["write_dotenv"] = bool(incoming.get("write_dotenv", True))
    for k in ("setup", "services_up", "services_down"):
        out[k] = str(incoming.get(k) or "").strip()
    from . import syspkgs
    out["system_packages"] = syspkgs.parse(incoming.get("system_packages"))  # raises ValueError on a bad name
    for c in incoming.get("checks") or []:
        cmd = str(c.get("command") or "").strip()
        if cmd:
            out["checks"].append({"command": cmd, "required": bool(c.get("required", True)), "name": str(c.get("name") or "").strip()[:80]})
    from .util import now
    out["repo"] = str(Path(repo).resolve())
    out["updated"] = now()
    ENV_DIR.mkdir(parents=True, exist_ok=True)
    write_json(_path(repo), out)
    try:
        _path(repo).chmod(0o600)
    except OSError:
        pass
    return out


def public(data: dict) -> dict:
    """What the browser may see: secret values masked, with a flag saying whether one is set."""
    pub = dict(data)
    pub["vars"] = [{**v, "value": MASK if v["secret"] and v["value"] else v["value"], "has_value": bool(v["value"])} for v in data["vars"]]
    pub["files"] = [{**f, "content": MASK if f.get("secret") and f["content"] else f["content"], "has_value": bool(f["content"])} for f in data["files"]]
    return pub


def parse_dotenv(text: str) -> list[dict]:
    """KEY=value lines from a pasted .env; names that look secret are marked secret."""
    rows = []
    for line in (text or "").splitlines():
        line = line.strip()
        if not line or line.startswith("#"):
            continue
        line = re.sub(r"^export\s+", "", line)
        if "=" not in line:
            continue
        name, value = line.split("=", 1)
        name, value = name.strip(), value.strip()
        if len(value) >= 2 and value[0] == value[-1] and value[0] in "\"'":
            value = value[1:-1]
        if _NAME.match(name):
            rows.append({"name": name, "value": value, "secret": looks_secret(name)})
    return rows


def looks_secret(name: str) -> bool:
    return bool(re.search(r"PASS|SECRET|TOKEN|KEY|PWD|CREDENTIAL|AUTH|COOKIE|TOTP|PRIVATE", name, re.I))


# ----------------------------------------------------------------------------- applying to a task
def env_vars(data: dict) -> dict:
    return {v["name"]: v["value"] for v in data["vars"]}


def for_path(path) -> dict:
    """The environment for whatever repository `path` (a worktree or repository) belongs to."""
    repo = source_repo(path)
    return load(repo) if repo else empty()


def secrets(data: dict) -> list[str]:
    vals = [v["value"] for v in data["vars"] if v["secret"] and len(v["value"]) >= 4]
    for f in data["files"]:
        if f.get("secret"):
            vals += [ln.split("=", 1)[1].strip().strip("\"'") for ln in f["content"].splitlines()
                     if "=" in ln and len(ln.split("=", 1)[1].strip().strip("\"'")) >= 6]
            # JSON/YAML style secrets: "key": "value" / key: value
            vals += [m.group(1) for m in re.finditer(r"[:=]\s*\"([^\"\s]{8,})\"", f["content"])]
    return sorted(set(vals), key=len, reverse=True)


def masker(data: dict):
    vals = secrets(data)
    if not vals:
        return lambda s: s
    rx = re.compile("|".join(re.escape(v) for v in vals))
    return lambda s: rx.sub(MASK, s) if isinstance(s, str) else s


def _exclude(wt: Path, patterns: list[str]) -> None:
    common = source_repo(wt)
    if not common:
        return
    info = common / ".git" / "info" if (common / ".git").is_dir() else common / "info"
    info.mkdir(parents=True, exist_ok=True)
    ex = info / "exclude"
    cur = ex.read_text(encoding="utf-8") if ex.exists() else ""
    lines = cur.splitlines()
    add = [f"/{p}" for p in patterns if f"/{p}" not in lines]
    if add:
        ex.write_text(cur + ("" if not cur or cur.endswith("\n") else "\n") + "\n".join(add) + "\n", encoding="utf-8")


def _tracked(wt: Path, rel: str) -> bool:
    return quiet(["git", "ls-files", "--error-unmatch", rel], cwd=wt, timeout=15).returncode == 0


def apply_to_worktree(wt, data: dict) -> list[str]:
    """Write the repository's files (and its .env) into a task worktree. Tracked files are never overwritten."""
    wt = Path(wt)
    written = []
    for f in data["files"]:
        rel = _clean_rel(f["path"])
        if _tracked(wt, rel):
            continue
        dest = wt / rel
        dest.parent.mkdir(parents=True, exist_ok=True)
        dest.write_text(f["content"], encoding="utf-8")
        written.append(rel)
    if data.get("write_dotenv", True) and data["vars"] and ".env" not in written and not _tracked(wt, ".env"):
        dest = wt / ".env"
        existing = dest.read_text(encoding="utf-8", errors="replace") if dest.exists() else ""
        # Keep lines the repository's own local .env had for names Relay does not set.
        ours = env_vars(data)
        kept = [ln for ln in existing.splitlines() if not re.match(rf"^\s*(?:export\s+)?({'|'.join(map(re.escape, ours))})\s*=", ln)]
        body = "\n".join(kept + [f"{k}={_quote(v)}" for k, v in ours.items()]).strip() + "\n"
        dest.write_text(body, encoding="utf-8")
        written.append(".env")
    if written:
        _exclude(wt, written)
    return written


def _quote(v: str) -> str:
    return f'"{v}"' if re.search(r"[\s#\"'$`\\]", v) else v


def describe(data: dict) -> str:
    """Names only, for agent prompts: agents learn what exists without ever seeing a value."""
    lines = []
    if data["vars"]:
        lines.append("- Environment variables set for this repository (already exported, and in `.env`): "
                     + ", ".join(f"`{v['name']}`" for v in data["vars"])
                     + ". Use them; never print their values, hard-code them or commit them.")
    if data["files"]:
        lines.append("- Local config files Relay placed in the worktree (git-ignored, never commit them): "
                     + ", ".join(f"`{f['path']}`" for f in data["files"]))
    if data.get("system_packages"):
        lines.append("- System packages Relay installed for this repository: " + ", ".join(f"`{p}`" for p in data["system_packages"])
                     + ". If another one is missing, record it as a blocked check naming the package; never run apt-get yourself.")
    if data.get("services_up"):
        lines.append(f"- Services Relay started for this task: `{data['services_up']}`.")
    if data["checks"]:
        lines.append("- The checks that prove the work (Relay runs them itself): "
                     + "; ".join(f"`{c['command']}`{'' if c['required'] else ' (optional)'}" for c in data["checks"]))
    return "\n".join(lines)

File: web/js/views/repoenv.js
// Repository environment editor: variables, local config files, setup, services and the checks that prove the work.
import { $, $$, esc, icon, toast, modal } from "../ui.js";
import { api } from "../api.js";
import { mountRepoConnectors } from "./connectors.js";

const MASK = "●●●●";
const secretName = (n) => /PASS|SECRET|TOKEN|KEY|PWD|CREDENTIAL|AUTH|COOKIE|TOTP|PRIVATE/i.test(n || "");

export async function openRepoEnv(repo, { onSaved } = {}) {
  let env;
  try { env = await api.repoEnv(repo.path); } catch (e) { toast("error", "Could not load the environment", e.message); return; }
  const varRow = (v = { name: "", value: "", secret: false }) => `<div class="renv-var">
      <input data-vn value="${esc(v.name)}" placeholder="NAME" spellcheck="false" autocomplete="off">
      <input data-vv type="${v.secret ? "password" : "text"}" value="${esc(v.value)}" placeholder="${v.secret && v.has_value ? "saved (hidden)" : "value"}" spellcheck="false" autocomplete="new-password">
      <label class="renv-secret" title="Hidden in the browser and masked in every log"><input type="checkbox" data-vs ${v.secret ? "checked" : ""}>secret</label>
      <button class="btn xs ghost" data-del title="Remove">${icon("x", "sm")}</button></div>`;
  const fileRow = (f = { path: "", content: "", secret: true }) => `<div class="renv-file">
      <div class="row" style="gap:6px"><input data-fp value="${esc(f.path)}" placeholder="relative/path, e.g. config/local.json" spellcheck="false" style="flex:1">
        <label class="renv-secret"><input type="checkbox" data-fs ${f.secret ? "checked" : ""}>secret</label>
        <button class="btn xs ghost" data-del title="Remove">${icon("x", "sm")}</button></div>
      <textarea data-fc rows="4" spellcheck="false" placeholder="${f.secret && f.has_value ? "saved (hidden); type to replace" : "file contents"}">${esc(f.secret && f.content === MASK ? "" : f.content)}</textarea></div>`;
  const checkRow = (c = { command: "", required: true }) => `<div class="renv-check">
      <input data-cc value="${esc(c.command)}" placeholder="python -m pytest -q" spellcheck="false">
      <label class="renv-secret"><input type="checkbox" data-cr ${c.required ? "checked" : ""}>required</label>
      <button class="btn xs ghost" data-del title="Remove">${icon("x", "sm")}</button></div>`;

  const m = modal(`<h2>${icon("shield")} Environment · ${esc(repo.name)}</h2>
    <p class="hint">Everything a task in this repository needs to run and test the software. It is stored on this server, never committed, applied to every new task's worktree, and secret values are masked in logs and hidden here after saving.</p>
    <div class="renv-sec"><div class="row between"><h3>Variables</h3><span class="row" style="gap:6px"><button class="btn xs" id="rePaste">${icon("file", "sm")}Paste .env</button><button class="btn xs" id="reAddVar">${icon("plus", "sm")}Add</button></span></div>
      <div id="reVars" class="stack" style="gap:6px">${env.vars.map(varRow).join("")}</div>
      <label class="row" style="gap:6px;margin-top:6px"><input type="checkbox" id="reDotenv" ${env.write_dotenv ? "checked" : ""}> Also write them to <code>.env</code> in the worktree (for apps that load dotenv files)</label></div>
    <div class="renv-sec"><div class="row between"><h3>Local files</h3><button class="btn xs" id="reAddFile">${icon("plus", "sm")}Add file</button></div>
      <p class="hint">Config files that are not in git: cookies, <code>config.local.json</code>, certificates.</p>
      <div id="reFiles" class="stack" style="gap:8px">${env.files.map(fileRow).join("")}</div></div>
    <div class="renv-sec"><h3>Setup and services</h3>
      <div class="field"><label>Setup command <span class="muted">(replaces the detected install)</span></label><input id="reSetup" value="${esc(env.setup)}" placeholder="leave empty to detect: npm ci, .venv + requirements + pytest, …" spellcheck="false"></div>
      <div class="field"><label>Start services before the team works</label><input id="reUp" value="${esc(env.services_up)}" placeholder="docker compose up -d db" spellcheck="false"></div>
      <div class="field"><label>Stop services when the task ends</label><input id="reDown" value="${esc(env.services_down)}" placeholder="docker compose down" spellcheck="false"></div></div>
    <div class="renv-sec"><div class="row between"><h3>Checks that prove the work</h3><button class="btn xs" id="reAddCheck">${icon("plus", "sm")}Add check</button></div>
      <p class="hint">Relay runs these itself after each work package. Required checks block delivery; optional ones are reported. Leave empty to detect them.</p>
      <div id="reChecks" class="stack" style="gap:6px">${env.checks.map(checkRow).join("")}</div></div>
    <div class="renv-sec" id="reConnectors"></div>
    <div class="modal-actions"><span class="muted" style="margin-right:auto">${env.updated ? `Saved ${esc(new Date(env.updated).toLocaleString())}` : "Not saved yet"}</span><button class="btn" data-close>Cancel</button><button class="btn primary" id="reSave">Save</button></div>`, { wide: true });

  const bindDel = () => $$("[data-del]", m.body).forEach((b) => (b.onclick = () => b.closest(".renv-var,.renv-file,.renv-check").remove()));
  const bindSecret = () => $$(".renv-var", m.body).forEach((r) => {
    const n = $("[data-vn]", r), v = $("[data-vv]", r), s = $("[data-vs]", r);
    s.onchange = () => { v.type = s.checked ? "password" : "text"; };
    n.onblur = () => { if (!n.dataset.touched && secretName(n.value) && !s.checked) { s.checked = true; v.type = "password"; } n.dataset.touched = "1"; };
  });
  const add = (id, html) => { $(id, m.body).insertAdjacentHTML("beforeend", html); bindDel(); bindSecret(); };
  bindDel(); bindSecret();
  mountRepoConnectors($("#reConnectors", m.body), repo);
  $("#reAddVar", m.body).onclick = () => { add("#reVars", varRow()); $$("[data-vn]", m.body).pop().focus(); };
  $("#reAddFile", m.body).onclick = () => add("#reFiles", fileRow());
  $("#reAddCheck", m.body).onclick = () => add("#reChecks", checkRow());
  $("#rePaste", m.body).onclick = () => {
    const p = modal(`<h2>Paste a .env file</h2><p class="hint">Lines like <code>NAME=value</code>. Names that look secret are marked secret. Existing variables with the same name are replaced.</p>
      <div class="field"><textarea id="peText" rows="12" spellcheck="false" placeholder="DATABASE=...\nPASSWORD=..."></textarea></div>
      <div class="modal-actions"><button class="btn" data-close>Cancel</button><button class="btn primary" id="peGo">Import</button></div>`, { wide: true });
    $("#peGo", p.body).onclick = async () => {
      try {
        const r = await api.repoEnvParse($("#peText", p.body).value);
        for (const v of r.vars) {
          const existing = $$(".renv-var", m.body).find((row) => $("[data-vn]", row).value.trim() === v.name);
          if (existing) existing.remove();
          add("#reVars", varRow(v));
        }
        toast("success", `Imported ${r.vars.length} variable(s)`, "Click Save to store them.");
        p.close();
      } catch (e) { toast("error", "Could not read that .env", e.message); }
    };
  };
  $("#reSave", m.body).onclick = async () => {
    const payload = {
      vars: $$(".renv-var", m.body).map((r) => ({ name: $("[data-vn]", r).value.trim(), value: $("[data-vv]", r).value, secret: $("[data-vs]", r).checked }))
        // A saved secret left untouched comes back as the mask, which keeps its stored value.
        .map((v) => { const old = env.vars.find((o) => o.name === v.name); return old && old.secret && old.has_value && v.value === "" ? { ...v, value: MASK } : v; })
        .filter((v) => v.name),
      files: $$(".renv-file", m.body).map((r) => ({ path: $("[data-fp]", r).value.trim(), content: $("[data-fc]", r).value, secret: $("[data-fs]", r).checked }))
        .map((f) => { const old = env.files.find((o) => o.path === f.path); return old && old.secret && old.has_value && f.content === "" ? { ...f, content: MASK } : f; })
        .filter((f) => f.path),
      write_dotenv: $("#reDotenv", m.body).checked,
      setup: $("#reSetup", m.body).value, services_up: $("#reUp", m.body).value, services_down: $("#reDown", m.body).value,
      checks: $$(".renv-check", m.body).map((r) => ({ command: $("[data-cc]", r).value.trim(), required: $("[data-cr]", r).checked })).filter((c) => c.command),
    };
    const btn = $("#reSave", m.body);
    btn.disabled = true;
    try {
      await api.saveRepoEnv(repo.path, payload);
      toast("success", `Environment saved for ${repo.name}`, "New tasks in this repository use it.");
      m.close();
      onSaved && onSaved();
    } catch (e) { toast("error", "Could not save", e.message); btn.disabled = false; }
  };
}
